cli/src/movement_util.rs

```diff
@@ -0,0 +1,217 @@
+
+    let target = match targets.as_slice() {
+        [target] => target,
+        [] => {
+            // We found no ancestor/descendant.
+            return Err(user_error(
+                direction.target_not_found_message(change_offset),
+            ));
+        }
+        commits => choose_commit(ui, workspace_command, &direction, commits)?,
+    };
+
+    Ok(target.clone())
+}
+
+fn choose_commit<'a>(
+    ui: &mut Ui,
+    workspace_command: &WorkspaceCommandHelper,
+    direction: &Direction,
+    commits: &'a [Commit],
+) -> Result<&'a Commit, CommandError> {
+    writeln!(
+        ui.stdout(),
+        "ambiguous {} commit, choose one to target:",
+        direction.cmd()
+    )?;
+    let mut formatter = ui.stdout_formatter();
+    let template = workspace_command.commit_summary_template();
+    let mut choices: Vec<String> = Default::default();
+    for (i, commit) in commits.iter().enumerate() {
+        write!(formatter, "{}: ", i + 1)?;
+        template.format(commit, formatter.as_mut())?;
+        writeln!(formatter)?;
+        choices.push(format!("{}", i + 1));
+    }
+    writeln!(formatter, "q: quit the prompt")?;
+    choices.push("q".to_string());
+    drop(formatter);
+
+    let choice = ui.prompt_choice(
+        "enter the index of the commit you want to target",
+        &choices,
+        None,
+    )?;
+    if choice == "q" {
+        return Err(user_error("ambiguous target commit"));
+    }
+
+    Ok(&commits[choice.parse::<usize>().unwrap() - 1])
+}
+
+pub fn move_to_commit(
+    ui: &mut Ui,
+    workspace_command: &mut WorkspaceCommandHelper,
+    direction: Direction,
+    edit: bool,
+    has_conflict: bool,
+    change_offset: u64,
+) -> Result<(), CommandError> {
+    let current_wc_id = workspace_command
+        .get_wc_commit_id()
+        .ok_or_else(|| user_error("This command requires a working copy"))?;
+    let edit = edit
+        || !&workspace_command
+            .repo()
+            .view()
+            .heads()
+            .contains(current_wc_id);
+    let target = get_target_commit(
+        ui,
+        workspace_command,
+        direction,
+        current_wc_id,
+        edit,
+        has_conflict,
+        change_offset,
+    )?;
+
+    let current_short = short_commit_hash(current_wc_id);
+    let target_short = short_commit_hash(target.id());
+    // We're editing, just move to the target commit.
+    if edit {
+        // We're editing, the target must be rewritable.
+        workspace_command.check_rewritable([target.id()])?;
+        let mut tx = workspace_command.start_transaction();
+        tx.edit(&target)?;
+        tx.finish(
+            ui,
+            format!("next: {current_short} -> editing {target_short}"),
+        )?;
+        return Ok(());
+    }
+    let mut tx = workspace_command.start_transaction();
+    // Move the working-copy commit to the new parent.
+    tx.check_out(&target)?;
+    tx.finish(ui, format!("next: {current_short} -> {target_short}"))?;
```

Comment:
the `next` in these two places should be `direction.cmd()`, right?

Reply:
Whoops! Fixed.